[Quant][fx] Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83263

Summary:

This change adds in input_type_to_index mappings to the backend patterns for `nn.functional.linear`, `nn.functional.conv1d`, `nn.functional.conv1d`, and `nn.functional.conv3d`.

This let's us remove `WEIGHT_INDEX_DICT` and `BIAS_INDEX_DICT` from `prepare.py`. 
Instead we pass around `backend_config` and check wether an arg is weight/bias agains that config


Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:
@andrewor14 

Subscribers:

Tasks:

Tags: quant, fx

Differential Revision: [D38705516](https://our.internmc.facebook.com/intern/diff/D38705516)

jcaip committed:
Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 reviewed (approved):
Looks great! Please add [Quant][fx] to the front of the PR title as well and fill out the PR description. For test plan you can just put something like:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

andrewor14 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -651,6 +687,8 @@ def maybe_insert_input_observers_for_node(
     prepare_custom_config: PrepareCustomConfig,
     backend_config: Optional[BackendConfig],
+    weight_index_dict: Dict[str, List[int]],
+    bias_index_dict: Dict[str, List[int]],
```
By the way, I know you just followed the TODO comment but one optional feedback I have is maybe instead of passing these `weight_index_dict` and `bias_index_dict` everywhere, we should just pass `backend_config` around. This would simplify the function signatures a little and avoid introducing yet another data structure that the user has to know about. Then we can just check against `backend_config.configs[pattern]._input_type_to_index` in `node_arg_is_weight` and `node_arg_is_bias`.

We don't have to do this in this PR though. Feel free to merge as is.

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[Quant][fx] Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

This change adds in input_type_to_index mappings to the backend patterns for `nn.functional.linear`, `nn.functional.conv1d`, `nn.functional.conv1d`, and `nn.functional.conv3d`.

This let's us remove `WEIGHT_INDEX_DICT` and `BIAS_INDEX_DICT` from `prepare.py`. 
Instead we pass around `backend_config` and check wether an arg is weight/bias agains that config


Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:
andrewor14 

Subscribers:

Tasks:

Tags: quant, fx

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -157,46 +154,40 @@ def is_activation_post_process_node(node: Node, modules: Dict[str, torch.nn.Modu
-                return True
+def node_arg_is_weight(node: Node, arg: Any, backend_config: BackendConfig) -> bool:
+    if isinstance(node, Node) and node.op == 'call_function':
+        weight_index = backend_config.configs[node.target]._input_type_to_index.get('weight')
```
I think you should also check if `node.target in backend_config.configs` (same in L167)

andrewor14 replied:
also nit: use double quotes to be consistent with other places

andrewor14 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -157,46 +154,40 @@ def is_activation_post_process_node(node: Node, modules: Dict[str, torch.nn.Modu
+def node_arg_is_weight(node: Node, arg: Any, backend_config: BackendConfig) -> bool:
+    if isinstance(node, Node) and node.op == 'call_function':
+        weight_index = backend_config.configs[node.target]._input_type_to_index.get('weight')
+        if weight_index and weight_index < len(node.args) and node.args[weight_index] is arg:
```
This won't work if `weight_index == 0` right? I think we need to explicitly check `weight_index is not None`

jcaip replied:
oh yeah good catch

andrewor14 commented:
Thanks @jcaip. This looks way cleaner than the previous version. Just left a few minor comments regarding the latest changes.

jcaip committed:
Update on "[Quant][fx] Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

This change adds in input_type_to_index mappings to the backend patterns for `nn.functional.linear`, `nn.functional.conv1d`, `nn.functional.conv1d`, and `nn.functional.conv3d`.

This let's us remove `WEIGHT_INDEX_DICT` and `BIAS_INDEX_DICT` from `prepare.py`. 
Instead we pass around `backend_config` and check wether an arg is weight/bias agains that config


Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:
andrewor14 

Subscribers:

Tasks:

Tags: quant, fx

[ghstack-poisoned]

jcaip committed:
Update on "[Quant][fx] Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT"


Summary:

This change adds in input_type_to_index mappings to the backend patterns for `nn.functional.linear`, `nn.functional.conv1d`, `nn.functional.conv1d`, and `nn.functional.conv3d`.

This let's us remove `WEIGHT_INDEX_DICT` and `BIAS_INDEX_DICT` from `prepare.py`. 
Instead we pass around `backend_config` and check wether an arg is weight/bias agains that config


Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:
andrewor14 

Subscribers:

Tasks:

Tags: quant, fx

[ghstack-poisoned]

jcaip commented:
@jcaip has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38705516).

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchmergebot merge

andrewor14 commented:
@jcaip I think this was already merged in [this commit](https://github.com/pytorch/pytorch/commit/d4bd88b64be3d7156cf8617e33faae8eca307ce4)? We can probably close this?

jerryzh168 commented:
thanks @jcaip for working on this, I think it's not fully done yet, since the end goal is to remove https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/utils.py#L57 and https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/utils.py#L69, but this is a good start. The configurations are not complete in backend_config currently I think, since there are more entries in the dictionary like layer_norm, group_norm etc.

jcaip commented:
@jerryzh168 Sounds good - I see that these dictionaries are also used in https://github.com/pytorch/pytorch/blob/aa405039545dc6f81b2b7d04e9d72434189cde1c/torch/ao/quantization/fx/_equalize.py and https://github.com/pytorch/pytorch/blob/13f42069a8659f24005e92a52432b9b91423150b/torch/ao/quantization/fx/convert.py. Should I pull them out of there too?
